primitives/state-machine/src/testing.rs

```diff
@@ -171,6 +194,31 @@ where
 		sp_externalities::set_and_run_with_externalities(&mut ext, execute)
 	}
 
+	/// Execute the given closure while `self`, with `proving_backend` as backend, is set as
+	/// externalities.
+	///
+	/// This implementation will wipe the proof recorded in between calls. Consecutive calls will
+	/// get their own proof from scratch.
+	///
+	/// To obtain a compact proof from the returned proof of this function, record the state root
+	/// before calling into this function, and use that as an argument to `into_compact_proof`.
+	pub fn execute_with_and_prove<'a, R>(
+		&mut self,
+		execute: impl FnOnce() -> R,
+	) -> (R, StorageProof) {
+		let tri_backend = self.backend.clone();
+		let proving_backend = InMemoryProvingBackend::new(&tri_backend);
+		let mut proving_ext = self.proving_ext(&proving_backend);
+		let outcome = sp_externalities::set_and_run_with_externalities(&mut proving_ext, execute);
+		let proof = proving_ext.backend.extract_proof();
+
+		// ensure that all changes are propagated, and the recorded is clean.
+		proving_ext.backend.clear_recorder();
```

Comment:
as per @cheme's previous feedback, the clear recorded is not really needed here. Will double check and remove.